app/public/static/js/components/ordered-puzzle.js

```diff
@@ -0,0 +1,196 @@
+            z: 0,
+          });
+        } else {
+          puzzlePiece.setAttribute("position", {
+            x: data.images[index].x * puzzlePieceWidth,
+            y: data.images[index].y * puzzlePieceHeight,
+            z: 0,
+          });
+        }
+
+        if (data.useTargets) {
+          // Create target
+          let textBoxProp = JSON.parse(JSON.stringify(data.images[index]));
+          textBoxProp.color = "yellow";
+          textBoxProp.width = puzzlePieceWidth;
+          textBoxProp.height = puzzlePieceHeight;
+          textBoxProp.xTarget = textBoxProp.xTarget * puzzlePieceWidth;
+          textBoxProp.yTarget = textBoxProp.yTarget * puzzlePieceHeight;
+
+          let target = document.createElement("a-entity");
+          target.setAttribute("id", "puzzle-target-" + index);
+          target.setAttribute(
+            "target-box",
+            "jsonData",
+            JSON.stringify(textBoxProp)
+          );
+          el.appendChild(target);
+
+          // Set image's target location
+          puzzlePiece.setAttribute(
+            "xTarget",
+            data.images[index].xTarget * puzzlePieceWidth
+          );
+          puzzlePiece.setAttribute(
+            "yTarget",
+            data.images[index].yTarget * puzzlePieceHeight
+          );
+          puzzlePiece.setAttribute("onTarget", false);
+
+          puzzlePiece.addEventListener("dragend", function (event) {
+            event.target.setAttribute("onTarget", isOnTarget(event.target));
+            if (isPuzzleComplete(puzzlePieceCache)) {
+              // emit event to update state
+              el.sceneEl.emit("solvedObject", { id: data.id });
+              // emit event to close popup (after 1 second for now) - TODO: disabled for now
+              // need designer input on whether we should autoclose this
+              // el.sceneEl.emit("dcc-success-close-popup", { seconds: 1 });
+              // emit event if it is last object, to indicate scene is solved
+              if (is_last_object) {
+                el.sceneEl.emit("dcc-success-scene-complete");
+              }
+            }
+          });
+        }
+
+        el.appendChild(puzzlePiece);
+      };
+      rawImageEl.src = data.images[i].imageSrc;
+    }
+
+    if (data.useTargets) {
+      // Once window is loaded, calculated if current puzzle pieces are on their target.
+      window.addEventListener("load", function () {
+        puzzlePieceCache.forEach((puzzlePiece) => {
+          puzzlePiece.setAttribute("onTarget", isOnTarget(puzzlePiece));
+        });
+      });
+    }
+  },
+  update: function () {
+    this.puzzleIsSolved = this.data.isSolved[this.id];
+
+    if (typeof this.puzzleIsSolved === "undefined") {
+      // not loaded yet, do nothing
+    } else if (this.puzzleIsSolved === true && this.useTargets) {
+      // Already solved
+      this.solvedPuzzleEntity.setAttribute("visible", "true");
+    } else {
+      // not solved yet, do nothing for now
+    }
```

Comment:
Nit: can just do the `else if` clause as the `if` and then both the not loaded and not solved can be caught with a single `else`

Reply:
Yeah I noticed that, I left it this way for the comments in the different cases; right now we don’t do anything specific but in the future we will :) so leaving it here for future references.
As an example, in the last else, we can randomize position instead of init/ that way, we only randomize if it hasn’t been solved. Of course this is for the future. I can change it for this PR you want

Reply:
Nah this is a good point -- good to leave as is.